tests/unit/classifier/Perceptron_unittest.cc

```diff
@@ -68,16 +72,69 @@ TEST(Perceptron, custom_hyperplane_initialization)
 	auto test_features = wrap(env->get_features_test());
 	auto test_labels = wrap(env->get_labels_test());
 
-	auto perceptron = some<CPerceptron>(features, labels);
-	perceptron->train();
+	auto perceptron = some<CPerceptron>();
+	perceptron->set_labels(labels);
+	perceptron->train(features);
 
 	auto weights = perceptron->get_w();
 
-	auto perceptron_initialized = some<CPerceptron>(features, labels);
+	auto perceptron_initialized = some<CPerceptron>();
 	perceptron_initialized->set_initialize_hyperplane(false);
 	perceptron_initialized->set_w(weights);
-	perceptron_initialized->set_max_iter(1);
+	perceptron_initialized->put<int32_t>("max_iterations", 1);
+	perceptron_initialized->set_labels(labels);
 
-	perceptron_initialized->train();
+	perceptron_initialized->train(features);
 	EXPECT_TRUE(perceptron_initialized->get_w().equals(weights));
 }
+
+TEST(Perceptron, prematurely_stop)
+{
+	auto env = linear_test_env->getBinaryLabelData();
+	auto features = wrap(env->get_features_train());
+	auto labels = wrap(env->get_labels_train());
+	auto test_features = wrap(env->get_features_test());
+	auto test_labels = wrap(env->get_labels_test());
+
+	auto perceptron = some<CPerceptron>();
+	perceptron->set_labels(labels);
+	perceptron->train(features);
+
+	auto results = wrap(perceptron->apply(test_features))
+	                   ->get<SGVector<float64_t>>("labels");
+
+	index_t iter = 0;
+
+	auto perceptron_stop = some<CPerceptron>();
+
+	std::function<bool()> callback = [&iter]() {
+		if (iter >= 1)
+		{
+			get_global_signal()->get_subscriber()->on_next(SG_BLOCK_COMP);
+			return true;
+		}
+		iter++;
+		return false;
+	};
+	perceptron_stop->set_callback(callback);
+
+	perceptron_stop->put<bool>("compute_bias", false);
+	perceptron_stop->set_labels(labels);
+	perceptron_stop->train(features);
+
+	auto results_itermediate = wrap(perceptron_stop->apply(test_features))
+	                               ->get<SGVector<float64_t>>("labels");
+
+	EXPECT_FALSE(results.equals(results_itermediate));
```

Comment:
I am confused why this would be the same: one perceptron was stopped after one iterations, and the other one was trained until convergence no?

Reply:
I train one perceptron for reference result then i trained another to stop after one iteration. the results_intermediate should not be equal to results. but when we continue_train we will converge properly and results_complete will be equal to results.

Reply:
I see. So this is about testing that one can stop a machine, then continue, and then the results match the process that didnt stop in the middle. OK!

BTW I think the `EXPECT_FALSE(results.equals(results_itermediate));` should be remove as this is not always necessarily the case. The last check is the one we care about.

You can check the status of the flags
m_converged == false
m_current_iteration == whatever

Reply:
pls name the test appropriately
stop_and_continue_same_as_no_stop or something :)